docs/enhancements/agent-upgrade.md

```diff
@@ -0,0 +1,189 @@
+with a newer version of the service.
+
+Note that in this context _version_ is really a container image name, like
+`registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141`.
+
+### Non-Goals
+
+Hosts will still need to be rebooted with a new version of the ISO if there
+are changes in the part of the agent that performs the initial registration.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster creator I want to be able to generate an ISO, boot and register
+hosts with a version of the service, then upgrade the service to a version that
+requires a newer version of the agent, and finally install the cluster without
+having to reboot the hosts.
+
+#### Story 2
+
+As a cluster Creator I want to be able to generate an ISO, boot and register
+hosts with a version of the service, then downgrade the service to a version
+that requires an older version of the agent, and finally install the cluster
+without having to reboot the hosts.
+
+### Implementation Details/Notes/Constraints
+
+A new `compatible-agent` validation will be added to the service to check if the
+agent is compatible.
+
+The logic for that validation will be a simple string comparison of the value
+of the `AGENT_DOCKER_IMAGE` environment variable with the value of the
+`discovery_agent_version` stored in the `hosts` table.
+
+To make this comparison useful it is important that the `AGENT_DOCKER_IMAGE`
+environment variable passed to the service doesn't use the `:latest` tag, as
+that renders the comparison useless. In the _SaaS_ stage environment that
+variable is already using a tag, so that shouldn't be a problem:
+
+```
+$ oc get deployment assisted-service -n assisted-installer-stage -o json | \
+jq '.spec.template.spec.containers[].env[] | select(.name == "AGENT_DOCKER_IMAGE")'
+{
+  "name": "AGENT_DOCKER_IMAGE",
+  "value": "registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141"
+}
+```
+
+In the integration and development environments we are using the `:latest` tag,
+which effectively disables this feature.
+
+The description of the `compatible-agent` validation will be the following:
+
+> The installation cannot start just yet because this host's agent is in the
+> process of being upgraded to a newer version, please wait, this could take
+> a few minutes.
+
+A new `upgrade-agent` step will be added to the service and to the agent.
+
+When the service receives a request for next steps from a host it will check if
+that host is in one of the `preparing-for-installation`, `installing` or
+`installing-in-progress` states. In it is then the service will return the next
+steps as usual. If it isn't then the service will retrieve from the database the
+result of the `compatible-agent` validation. If the validation passed then the
+service will return the next steps as usual, otherwise it will replace all the
+usual steps the `upgrade-agent` step.
```

Comment:
I think that if the `compatible-agent` validation failed the host status should be `insufficient`.
So in essence we can check the `compatible-agent` validation for sending the `agent-upgrade` step only if the host state is `insufficient`.

Reply:
@eranco74 I got confused. the first sentence implies you do the validation first, and then move the status to insufficient.
The second one implies that you do the test only if the status is insufficient.
Or there are two kind of validations - one which just moves the status to insufficient, and the other is actually requesting for an upgrade to take place?

Reply:
Validation should run on all states except "installation ones". We should be able to move from Ready to insufficient if agent image was changed, we have pretty much examples of it already

Reply:
I am keeping the current description, if you disagree please reopen this conversation.